### PR TITLE
Optional KL Divergence Computation

### PR DESCRIPTION
**Pull request by robbietuk**
Addresses #697 

Combines the objective function value computation to run through the same code as the gradient computation. For numerical computational reasons, when computing a single function/gradient evaluation (i.e. a single bin) an upper and lower threshold are appiled to prevent singlarities (divide by ~zero) and negative values. Both the functional and gradient values should be subject to the same bounds and therefore this PR combines these evaluations into `poisson_divide_and_truncate`. 

If a `log_likelihood_ptr` is parsed, the function will compute the function. Likewise, if the `do_division == true` the function will compute the gradient. 

Adds bool `use_KL_divergence` that is parsed from `PoissonLogLikelihoodWithLinearModelForMean` methods through to a set of functions in `recon_array_functions.cxx`. `use_KL_divergence` can be set at objective function level and by default this is off, so backward compatable. 

- [x] Test with evaluation of objective functions (Compare STIR master, and branch LL and KL computations). 

STIR master and branch LL are identical. KL produces value ~1000x smaller than LL.

- [ ] Documentation: Release notes, Users guide.
- [ ] Check mismatch between STIR master LL and this PR LL values (https://github.com/UCL/STIR/pull/760#issuecomment-780743088)

- [ ] Test MPI Code

### CODE REVIEW

File: src/buildblock/recon_array_functions.cxx [hunk 1]
```diff
@@ -224,48 +225,50 @@ void divide_and_truncate(Viewgram<float>& numerator,
...
+          // Note that this includes the case if a negative denominator
+          // (in case somebody forward projects an image with negatives)
+          if (numerator[r][b] > max_quotient*denominator[r][b])
+          {
+            // cancel singularity
+            count++;
+            if (log_likelihood_ptr != NULL)
+            {
+              // Substitute denominator[r][b] (i.e. ybar) for num / max_quotient (i.e. y/ max_quotient)
+              sub_result += compute_data_fit_term(numerator[r][b], numerator[r][b] / max_quotient, use_KL_divergence, small_value);
```
Comment (robbietuk) [inline src/buildblock/recon_array_functions.cxx:257]: This `compute_data_fit_term` call is different to how `accumulate_loglikelihood` handles it.

File: src/buildblock/recon_array_functions.cxx [hunk 2]
```diff
@@ -344,12 +349,13 @@ void divide_array(DiscretisedDensity<3,float>& numerator, const DiscretisedDensi
     
 }
 
-//  MJ 03/01/2000 for loglikelihood computation
+//  MJ 03/01/2000 for loglikelizhood computation
```
Comment (robbietuk) [inline src/buildblock/recon_array_functions.cxx:352]: typo?

File: src/include/stir/recon_array_functions.h [hunk 3]
```diff
@@ -111,8 +113,10 @@ void divide_array(DiscretisedDensity<3,float>& numerator, const DiscretisedDensi
 void accumulate_loglikelihood(Viewgram<float>& projection_data, 
 			 const Viewgram<float>& estimated_projections,
 			 const int rim_truncation_sino,
-			 double* accum);
+			 double* accum,
+			 const bool use_KL_divergence);
 
+float compute_data_fit_term(float y, float ybar, bool use_KL_divergence, float small_value);
```
Comment (robbietuk) [inline src/include/stir/recon_array_functions.h:119]: Needs documentation

File: src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMean.h [hunk 4]
```diff
@@ -237,12 +237,17 @@ public  GeneralisedObjectiveFunction<TargetT>
   void 
     fill_nonidentifiable_target_parameters(TargetT& target, const float value ) const;
 
+  bool get_use_KL_divergence();
+
+  void set_use_KL_divergence(const bool KL);
```
Comment (robbietuk) [inline src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMean.h:242]: Needs documentation

File: src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMean.h [hunk 5]
```diff
@@ -237,12 +237,17 @@ public  GeneralisedObjectiveFunction<TargetT>
...
+
  private:
 
   std::string sensitivity_filename;
   std::string subsensitivity_filenames;
   bool recompute_sensitivity;
   bool use_subset_sensitivities;
+  bool use_KL_divergence;
```
Comment (robbietuk) [inline src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMean.h:250]: Documentation?

File: src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx [hunk 6]
```diff
@@ -1019,23 +1025,23 @@ void distributable_sensitivity_computation(
                                     start_time_of_frame,
                                     end_time_of_frame,
                                     &RPC_process_related_viewgrams_sensitivity_computation,
-                                    caching_info_ptr
+                                    caching_info_ptr,
+                                    false
```
Comment (robbietuk) [inline src/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMeanAndProjData.cxx:1029]: Pass `use_KL_divergence` as false here. This is in sensitivity computation. No need to use it but distributable computation needs the bool? Could pass `NULL` instead?

**Review by robbietuk — COMMENTED** [review]
Prior to this PR, no `RPC_process_related_viewgrams_gradient` method utilised the log-likelihood computation. As mentioned https://github.com/UCL/STIR/issues/697#issuecomment-734802568, the computation in `divide_and_truncate` was wrong anyway. 

`divide_and_truncate` is a little complex of a function as it does both the division and computation (`if log_likelihood_ptr != NULL`).

File: src/buildblock/recon_array_functions.cxx (hunk 2 above)
Comment (KrisThielemans) [reply src/buildblock/recon_array_functions.cxx:352]: edit per accident I guess. revert

File: src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMean.h (hunk 5 above)
Comment (KrisThielemans) [reply src/include/stir/recon_buildblock/PoissonLogLikelihoodWithLinearModelForMean.h:250]: `private` members. as you like. could just refer `\see set_KL_divergence`

File: src/buildblock/recon_array_functions.cxx [hunk 7]
```diff
@@ -224,48 +225,50 @@ void divide_and_truncate(Viewgram<float>& numerator,
       };
 
 #else
-      if(b<bs+rim_truncation_sino ||
-	 b>be-rim_truncation_sino ) 
-	{
-	  numerator[r][b] = 0;
-	}
+      if(b<bs+rim_truncation_sino || b>be-rim_truncation_sino )
```
Comment (KrisThielemans) [inline src/buildblock/recon_array_functions.cxx:228]: please refrain from reformatting code that isn't related to this PR

File: src/buildblock/recon_array_functions.cxx [hunk 8]
```diff
@@ -224,48 +225,50 @@ void divide_and_truncate(Viewgram<float>& numerator,
...
-		}
-	      else
-		{
-		  if (log_likelihood_ptr != NULL) 
-		    sub_result -= double(num*log(denom));
-		  num = num/denom;
-		}
-	    }
-	}
+      {
```
Comment (KrisThielemans) [inline src/buildblock/recon_array_functions.cxx:231]: a lot of this is just reformatting as well. I can see why, but this code is at the heart of a lot of things. I'd rather not change it unless it needs to

File: src/buildblock/recon_array_functions.cxx (hunk 1 above)
Comment (KrisThielemans) [reply src/buildblock/recon_array_functions.cxx:257]: I haven't checked, but obviously we should have only one way/function to do things (or at least make sure the result is the same)

File: src/buildblock/recon_array_functions.cxx [hunk 9]
```diff
@@ -366,35 +372,44 @@ void accumulate_loglikelihood(Viewgram<float>& projection_data,
...
-	  if (projection_data[r][b]<=small_value)
-	    sub_result += - double(new_estimate);
-	  else
-	    sub_result += projection_data[r][b]*log(double(new_estimate)) - double(new_estimate);
-	}
+      if(!( b<bs+rim_truncation_sino || b>be-rim_truncation_sino ))
+      {
+        // if (estimated_projections[r][b] == 0)
+        //  std::cerr << "Zero at " << r << ", " << b <<'\n';
+        const float new_estimate = max(estimated_projections[r][b], projection_data[r][b]/max_quotient);
```
Comment (KrisThielemans) [inline src/buildblock/recon_array_functions.cxx:388]: all this seems reformatting

File: src/buildblock/recon_array_functions.cxx [hunk 10]
```diff
@@ -366,35 +372,44 @@ void accumulate_loglikelihood(Viewgram<float>& projection_data,
...
+        sub_result += compute_data_fit_term(projection_data[r][b], new_estimate, use_KL_divergence, small_value);
```
Comment (KrisThielemans) [inline src/buildblock/recon_array_functions.cxx:389]: I like putting this in a function, such that we can call it. I'm not convinced about the name though. At some point, we want to have least squares data-fit as well. Any ideas? (`compute_Poisson_data_fit`?)

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
I suspect you'll need to make some small adjustments in the MPI code as well. Currently not tested on Travis. sorry. I can run the test if you like.

File: src/buildblock/recon_array_functions.cxx (hunk 10 above)
Comment (robbietuk) [reply src/buildblock/recon_array_functions.cxx:389]: I am bad at naming things and this is a bad name. `compute_Poisson_data_fit` is good. Ill keep thinking.

File: src/buildblock/recon_array_functions.cxx (hunk 1 above)
Comment (robbietuk) [reply src/buildblock/recon_array_functions.cxx:257]: In that case this should be 
```
sub_result += compute_data_fit_term(numerator[r][b], denominator[r][b], use_KL_divergence, small_value);
```
regardless of `if (numerator[r][b] > max_quotient*denominator[r][b])`. Log-likelihood and KL divergence should not have special cases for divide by zeros?

**Comment by robbietuk** [comment]
Objective function evaluations (LogLikelihood):
 - STIR master = 7845451546.05617
 - This PR = 7845451547.569616
   - Diff = -1.513446

To me this indicates that there was a mismatch between the function and gradient due to thresholding. I will need to investigate further.

File: src/buildblock/recon_array_functions.cxx [hunk 11]
```diff
@@ -231,41 +236,49 @@ void divide_and_truncate(Viewgram<float>& numerator,
...
-	    {
-	      const float max_quotient = 10000.F;
-	      const float denom = denominator[r][b];
+          numerator[r][b] = 0;
+          if (numerator[r][b]<0)
+            count2++;
+        }
+        else
+        {
+          const float max_quotient = 10000.F;
```
Comment (robbietuk) [inline src/buildblock/recon_array_functions.cxx:251]: Can remove white space changes

File: src/buildblock/recon_array_functions.cxx [hunk 12]
```diff
@@ -346,54 +361,79 @@ void divide_array(DiscretisedDensity<3,float>& numerator, const DiscretisedDensi
...
-  
+#if 1
+  int count = 0;
+  int count2 = 0;
+      poisson_divide_and_truncate(projection_data,
+                                  estimated_projections,
+                                  rim_truncation_sino,
+                                  count, count2,
+                                  log_likelihood_ptr, use_KL_divergence, false);
+
```
Comment (robbietuk) [inline src/buildblock/recon_array_functions.cxx:380]: `accumulate_loglikelihood` now uses the `poisson_divide_and_truncate` method with a `log_likelihood_ptr` and `do_division=false`

File: src/buildblock/recon_array_functions.cxx [hunk 13]
```diff
@@ -346,54 +361,79 @@ void divide_array(DiscretisedDensity<3,float>& numerator, const DiscretisedDensi
...
-	    max(estimated_projections[r][b], 
-		projection_data[r][b]/max_quotient);
-	  if (projection_data[r][b]<=small_value)
-	    sub_result += - double(new_estimate);
-	  else
-	    sub_result += projection_data[r][b]*log(double(new_estimate)) - double(new_estimate);
-	}
+        const float new_estimate = max(estimated_projections[r][b], projection_data[r][b]/max_quotient);
+        sub_result += compute_Poisson_data_fit(projection_data[r][b], new_estimate, use_KL_divergence, small_value);
+      }
```
Comment (robbietuk) [inline src/buildblock/recon_array_functions.cxx:413]: Can possibly undo this change if we keep the `#IF #ELSE #ENDIF` code.

File: src/include/stir/recon_array_functions.h [hunk 14]
```diff
@@ -111,8 +113,10 @@ void divide_array(DiscretisedDensity<3,float>& numerator, const DiscretisedDensi
 void accumulate_loglikelihood(Viewgram<float>& projection_data, 
 			 const Viewgram<float>& estimated_projections,
 			 const int rim_truncation_sino,
-			 double* accum);
+			 double* accum,
+			 const bool use_KL_divergence);
 
+float compute_Poisson_data_fit(float y, float ybar, bool use_KL_divergence, float lower_threshold);
```
Comment (robbietuk) [inline src/include/stir/recon_array_functions.h:119]: Need documentation

**Comment by robbietuk** [comment]
I think I will abandon this PR. The idea is great but the implementation is complex.
